python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py

```diff
@@ -400,17 +406,72 @@ def select(self, collection: CollectionRecord,
+        if run is not None:
+            if collection.type is CollectionType.RUN:
+                if run is SimpleQuery.Select:
+                    # If the collection we're searching is a RUN, we know that
+                    # if we find the dataset in that collection, then that's
+                    # the datasets's run; we don't need to query for it.
+                    query.columns.append(sqlalchemy.sql.literal(collection.key).label(self._runKeyColumn))
+                elif run != collection.name:
+                    # This [sub]query is doomed to yield no results; dataset
+                    # cannot be in more than one run.
+                    return None
+                else:
+                    query.where.append(self._static.dataset.columns[self._runKeyColumn] == collection.key)
+            else:
+                static_kwargs[self._runKeyColumn] = (
+                    SimpleQuery.Select if run is SimpleQuery.Select else self._collections.find(run).key
+                )
+                need_static_table = True
+        # Ingest date can only come from the static table.
+        if ingestDate is not None:
+            need_static_table = True
+            if ingestDate is SimpleQuery.Select:
+                static_kwargs["ingest_date"] = SimpleQuery.Select
+            else:
+                assert isinstance(ingestDate, Timespan)
+                # Timespan is astropy Time (usually in TAI) and ingest_date is
+                # TIMESTAMP, convert values to Python datetime for sqlalchemy.
+                if ingestDate.isEmpty():
+                    raise RuntimeError("Empty timespan constraint provided for ingest_date.")
+                if ingestDate.begin is not None:
+                    begin = ingestDate.begin.utc.datetime  # type: ignore
+                    query.where.append(self._static.dataset.columns.ingest_date >= begin)
+                if ingestDate.end is not None:
+                    end = ingestDate.end.utc.datetime  # type: ignore
+                    query.where.append(self._static.dataset.columns.ingest_date < end)
+        # If we need the static table, join it in via dataset_id and
+        # dataset_type_id
+        if need_static_table:
             query.join(
-                self._tags,
-                onclause=(self._static.dataset.columns.id == self._tags.columns.dataset_id),
-                **kwargs
+                self._static.dataset,
+                onclause=(dataset_id_col == self._static.dataset.columns.id),
+                **static_kwargs,
             )
+            # Also constrain dataset_type_id in static table in case that helps
+            # generate a better plan.
+            # We could also include this in the JOIN ON clause, but my guess is
+            # that that's a good idea IFF it's in the foreign key, and right
+            # now it isn't.
+            query.where.append(self._static.dataset.columns.dataset_type_id == self._dataset_type_id)
```

Comment:
I don't think that this is helping as the constraint on dataset_id is already very strict, but I hope it is not hurting either.